src/core/service.c

```diff
@@ -1278,15 +1277,17 @@ static void service_set_state(Service *s, ServiceState state) {
         if (!IN_SET(state, SERVICE_START_POST, SERVICE_RUNNING, SERVICE_RELOAD, SERVICE_RELOAD_SIGNAL, SERVICE_RELOAD_NOTIFY))
                 service_stop_watchdog(s);
 
-        /* For the inactive states unit_notify() will trim the cgroup,
-         * but for exit we have to do that ourselves... */
-        if (state == SERVICE_EXITED && !MANAGER_IS_RELOADING(UNIT(s)->manager))
-                unit_prune_cgroup(UNIT(s));
+        /* For the inactive states unit_notify() will trim the cgroup, and service_enter_dead() will
+         * destroy runtime data. But for exit we have to do that ourselves... */
+        if (state == SERVICE_EXITED && !MANAGER_IS_RELOADING(u->manager)) {
+                unit_prune_cgroup(u);
+                unit_destroy_runtime_data(u, &s->exec_context);
```

Comment:
hmm, this might be ok to do, but with some restrictions. i.e. if ExecStop=/ExecStopPost= is used we probably shouldn't do this, so that runtime stuff can stay around